Add post for common pitfalls in the SMC integration (#304)
openSUSE/obs-landing: _posts/development/2021-07-12-scm-integration-pitfalls.md

### Non-existent Project or Package

**Error message:** "Project not found <project>" and "Package not found <package>".

In the configuration file `.obs/workflows.yml`, you have to specify the name of the project and package you want to connect with the SCM.

Make sure the project and package really exist in OBS before you set up the webhook.

### Missing or Invalid `_service` File

The OBS package you are going to use in the integration should contain a `_service` file pointing to the right SCM repository and branch, as it's described 
in the [Use the SCM Sources in Your OBS package](/2021/05/31/scm-integration/) section of a previous blog post. If the `_service` file is not setup correctly 
the package will still be able to build but it will not contain the latest changes from your target branch.

You can check a sample of `_service` file published on [OBS](https://build.opensuse.org/package/view_file/OBS:Server:Unstable/obs-server/_service?expand=0&rev=424295e96102d316454e624b0a3d1ae3).

> ### No build result updates arrive your PR/MR

Reviewer: To fix the capitalization and a missing `In`:
```suggestion
### No Build Result Updates Are Displayed in Your PR/MR
```

Reviewer 2: Done.

> If a package can't be build because of one of the many reasons a package in the OBS can't be build, then the status of the check in your pull/merge request will never change. Some of the more frequent causes of packages not building are:

Reviewer: ```suggestion
If somehow a package can't be build, then the status of the check in your pull/merge request will never change. Some of the more frequent causes of packages not building are:
```

Reviewer 2: Done.

---
layout: post
title: Common Pitfalls to Avoid in the Continuous Integration with OBS and GitHub/GitLab
category: development
---

Many of you are already playing around with our new beta feature: Source Code Management (SCM) system integration with OBS. You might have faced some errors during this time.

{% include partials/_series-of-posts-about-scm-integration.md %}

Here you can find a list of the most common pitfalls you can face and how to solve them.

> ## Where to See the Errors?

Reviewer 3: Okay only one thing missing to make this more readable IMHO. Put this section to the bottom and add a sentence like:

`Some errors you will also see in your pull/merge request or in OBS itself.`

Reviewer 2: Done.

### Non-existent Project or Package

**Error message:** "Project not found <project>" and "Package not found <package>".

In the configuration file `.obs/workflows.yml`, you have to specify the name of the project and package you want to connect with the SCM.

Make sure the project and package really exist in OBS before you set up the webhook.

### Missing or Invalid `_service` File

The OBS package you are going to use in the integration should contain a `_service` file pointing to the right SCM repository and branch, as it's described 
in the [Use the SCM Sources in Your OBS package](/2021/05/31/scm-integration/) section of a previous blog post. If the `_service` file is not setup correctly 
the package will still be able to build but it will not contain the latest changes from your target branch.

You can check a sample of `_service` file published on [OBS](https://build.opensuse.org/package/view_file/OBS:Server:Unstable/obs-server/_service?expand=0&rev=424295e96102d316454e624b0a3d1ae3).

> ### No Repositories Assigned to the Project
> 
> Don't forget to assign the repositories and architectures you want your package to get build for.
> You can add them in the Repositories tab of the package's project page.

Reviewer 3: ```suggestion
### No build result updates arrive your PR/MR

If a package can't be build for any of the many reasons a package in the OBS can't be build then the status of the check in your pull/merge request will never change. Some of the more frequent ones are:

- A build is unresolvale
- A build is broken
- No repositories defined on your source project

You should check the project directly on OBS and [help each other](/support) with problems.
```

Author: @hennevogel applied, thanks!

---
layout: post
title: Common Pitfalls to Avoid in the Continuous Integration with OBS and GitHub/GitLab
category: development
---

Many of you are already playing around with our new beta feature: Source Code Management (SCM) system integration with OBS. You might have faced some errors during this time.
Here you can find a list of the most common pitfalls you can face and how to solve them.

{% include partials/_series-of-posts-about-scm-integration.md %}

## Where to See the Errors?

Once you set the webhooks, you can detect the errors returned by OBS in the webhook response.

In GitHub, you can find them in:
Settings -> Webhooks -> Edit webhook -> **Recent Deliveries** (tab)

<figure>
  <img src="/images/posts/document-scm/github-error-reporting.png" alt="Webhook errors" width="1000px" />

>   <figcaption>Github Webhook Errors</figcaption>

Reviewer: ```suggestion
  <figcaption>GitHub Webhook Errors</figcaption>
```

</figure>

In GitLab, you can find them under:
Project -> Settings -> Webhooks -> Edit -> **Recent Deliveries** (at the bottom of the page).

<figure>
  <img src="/images/posts/document-scm/gitlab-error-reporting.png" alt="Webhook errors" width="1000px" />

>   <figcaption>Gitlab Webhook Errors</figcaption>

Reviewer: ```suggestion
  <figcaption>GitLab Webhook Errors</figcaption>
```

</figure>

## Pitfalls & Solutions

> Here you have a list of common pitfalls and their possible solutions.

Reviewer: It reads better and it's shorter. I removed possible because it sounds like what we wrote might not solve the issues, although we're pretty sure it is.
```suggestion
Here's a list of common pitfalls and their solutions.
```

### Not Authorized

> **Error message:** *"You are not authorized to create this Token/workflow."*

Reviewer 3: This is not going to be in the webhook request cycle right?

Reviewer 2: If I remember correctly, this is the exact error received in the webhook.

Reviewer 3: okay 👍🏿

> This SCM integration feature is only available for users in Beta Program. Please join the Beta Program in OBS to be able to use this new feature.

Reviewer: ```suggestion
This SCM integration feature is only available for users in the beta program. Please [join the beta program](/2018/10/04/the-beta-program/) in OBS to be able to use this new feature.
```

> See how [here](https://openbuildservice.org/2018/10/04/the-beta-program/).

Reviewer: This sentence isn't needed anymore now that we have the `join the beta program` link in the previous sentence.
```suggestion
```

### Non-existent YAML Configuration File

**Error message:** ".obs/workflows.yml could not be downloaded on the SCM branch <branch_name>".

> OBS assumes that your repository contains a `.obs/workflows.yml` file. OBS only considers this file to be present in your default branch (i.e master or main) and will not consider any updates to this file in other branches.

Reviewer 3: ```suggestion
The feature requires that your SCM repository contains a `.obs/workflows.yml` file. OBS only considers this file in your default branch (i.e master or main) and will not consider this file in other branches or forks.
```

Reviewer 3: I will see the same error if the file isn't accessible for any other reasons right?

Reviewer 2: Yes, the error message will be the same.

Reviewer 3: You should mention this.

Make sure you place the configuration file in the default branch before you set up your webhooks.

### Invalid YAML Configuration File

**Error message:** *"Invalid workflow step definition"*, *"Unsupported filters"* or *"Filters <filter_names> have unsupported keys."*

> To get the idea of how to format the `.obs/workflows.yml` file you can see the sample [here](https://github.com/openSUSE/open-build-service/tree/master/.obs). You can find some guidelines in this article:

Reviewer: Let's point directly to the file since this is what users want to see:
```suggestion
To get an idea of how to format the `.obs/workflows.yml` file, you can see a sample [here](https://github.com/openSUSE/open-build-service/tree/master/.obs/workflows.yml). You can find some guidelines in this article:
```

> [Continuous Integration with OBS and GitHub/GitLab](https://openbuildservice.org/2021/05/31/scm-integration/).

Reviewer: Please use relative URLs for all links across the blog post:
```suggestion
[Continuous Integration with OBS and GitHub/GitLab](/2021/05/31/scm-integration/).
```

 [YAML Checker](https://yamlchecker.com) can help you too.

### SCM Token Not Provided

**Error message:** *"Bad credentials".*
  

> Don't forget to provide your SCM personal access token when you create an OBS workflow token.

Reviewer 3: You can't forget this, it's validated. If you get this error, it's invalid credentials. What do I have to do to get out of this situation?

Reviewer 2: ATM the only option the user has to create a token is using OSC. Users can create a new token but once we have UI options available for users where they can edit their token, that will be easier and users can replace invalid SMC tokens with the valid ones.

Reviewer 3: But we don't so you should tell people that they will have to delete the old one, create a new one with good credentials.

Reviewer 2: Okay, sure.

Check sections 1 and 2 of [Continuous Integration with OBS and GitHub/GitLab](https://openbuildservice.org/2021/05/31/scm-integration/) for more details.

### Non-existent Project or Package

**Error message:** *"Project not found <project>"* and *"Package not found <package>".*

> In the configuration file, `.obs/workflows.yml`, you have to specify the name of the project and package you want to connect with the SCM.

Reviewer: Unneeded comma:
```suggestion
In the configuration file `.obs/workflows.yml`, you have to specify the name of the project and package you want to connect with the SCM.
```

Make sure the project and package really exist in OBS before you set up the webhook.

### The Package Doesn't Build

When everything seems to be correctly set up and you are waiting for the checks in your SCM to get updated
but this never happens, these are the possible problems:

#### Missing or Invalid `_service` File

The OBS package you are going to use in the integration should contain a `_service` file pointing to the right SCM repository and branch, as it's described 
in the [Use the SCM Sources in Your OBS package](https://openbuildservice.org/2021/05/31/scm-integration/) section of a previous blog post.

> You can check a sample of `_service` file published on [OBS](https://build.opensuse.org/package/view_file/OBS:Server:Unstable/obs-server/_service?expand=0&rev=424295e96102d316454e624b0a3d1ae3)

Reviewer 3: This makes it sounds as if reporting depends on using obs-service-tar_scm. But packages without a _service file will be branched, build and the result will be reported.

Reviewer: Sure, OBS will still build and report, but the build won't include the source changes if a `_service` file isn't present. Workflows are pretty useless without this.

Reviewer 3: You are right maybe, depends a bit on your package and the services includes, but that's not the point, correct?

The point is of this section is: The Package Doesn't Build. While it does build (but not the things you would maybe expect it to build).

Reviewer 2: Alright, I've updated that section.

#### No Repositories Assigned to the Project

Don't forget to assign the repositories and architectures you want your package to get build for.

> You can add them in the Repositories tab of the package's project page.

Reviewer 3: A package can not build for many more reasons. Maybe something generic like your package never built? And how do I get out of this situation?

Author: This section is no-longer part of `The Package Doesn't Build`, this is now an independent section.
This is the problem we faced in the manual tests. We can always add new building problems and solutions in the future.

Is it ok for you if we leave it like it is?

@hennevogel

Reviewer 3: It's very confusing IMHO. Now so even more because I don't know which project/package you are talking about. The source or the target?

Reviewer 3: Made a suggestion
